cachi2/core/package_managers/pip.py

```diff
@@ -1476,45 +1477,46 @@ def _split_hashes_from_options(options: list[str]) -> tuple[list[str], list[str]
 def _process_req(
     req: PipRequirement,
     requirements_file: PipRequirementsFile,
-    require_hashes: bool,
     pip_deps_dir: RootedPath,
     download_info: dict[str, Any],
     dpi: Optional[DistributionPackageInfo] = None,
 ) -> dict[str, Any]:
     download_info["kind"] = req.kind
     download_info["requirement_file"] = str(requirements_file.file_path.subpath_from_root)
-    download_info["checksum_matched"] = False
+    download_info["missing_req_file_checksum"] = True
+    # "package_type" is *only* needed for PyPI deps
+    download_info["package_type"] = ""
 
-    def _checksum_match(path: Path, checksum_info: Iterable[ChecksumInfo]) -> bool:
-        matched: bool = False
+    def _checksum_must_match_or_path_unlink(
+        path: Path, checksum_info: Iterable[ChecksumInfo]
+    ) -> None:
         try:
             # returns None, raises PackageRejected on failure
             must_match_any_checksum(path, checksum_info)
-            matched = True
         except PackageRejected:
             path.unlink()
             log.warning("Download '%s' was removed from the output directory", path.name)
-        return matched
 
     if dpi:
-        if dpi.has_checksums_to_match():
-            download_info["checksum_matched"] = _checksum_match(dpi.path, dpi.checksums_to_match)
-
+        if dpi.req_file_checksums:
+            download_info["missing_req_file_checksum"] = False
+        if dpi.has_checksums_to_match:
+            _checksum_must_match_or_path_unlink(dpi.path, dpi.checksums_to_match)
```

Comment:
Just checking my understanding: if the package has any `req_file_checksum`, we're marking it as having checksums in `requirements.txt` . Then, it will always pass the `if dpi.has_checksums_to_match` (since it has at least one `req_file_checksum`), and in case the checksum does not match, the package is discarded.

Could there be a scenario in which the `req_file_checksum` did not match, but a `pypi_checksum` matched, and we allowed the package to exist in the prefetched content? I think if this happened, the SBOM would lead us to believe that there was nothing wrong with the package, and that its checksums were properly matched.

Reply:
> if the package has any req_file_checksum, we're marking it as having checksums in requirements.txt

If this specific *download*, that is to say **this** artifact, has a checksum in the *current* requirements file.

> Could there be a scenario in which the req_file_checksum did not match

I'll have to think about it for a bit and get back to you

Reply:
If we:
- download an artifact
- calculate its checksum
- the checksum is correct
- we *have* `req_file_checksums` AND there's no match with `pypi_checksum`

then `checksums_to_match` *should be* empty, because `self.req_file_checksums.intersection(self.pypi_checksum)` will be empty. If `checksums_to_match` is empty, the check for that download fails at some point.

So, I think the answer to your question is that we *might* download something we "shouldn't", but the checksum matching will fail later on, so we should be OK

Reply:
You're right. We actually don't even download it: https://github.com/containerbuildsystem/cachi2/blob/main/cachi2/core/package_managers/pip.py#L1912-L1918